tools/jvm/src/main/scala/org/scalajs/core/tools/optimizer/ScalaJSClosureOptimizer.scala

```diff
@@ -85,28 +85,44 @@ class ScalaJSClosureOptimizer {
     }
 
     logTime(logger, "Closure: Write result") {
-      writeResult(result, compiler, cfg.output)
+      writeResult(result, compiler, cfg)
     }
   }
 
   private def writeResult(result: Result, compiler: ClosureCompiler,
-      output: WritableVirtualJSFile): Unit = {
+      cfg: Config): Unit = {
+    val output = cfg.output
 
-    val outputContent = if (result.errors.nonEmpty) ""
-    else "(function(){'use strict';" + compiler.toSource + "}).call(this);\n"
+    val hasErrors = result.errors.nonEmpty
+
+    def withNewLine(str: String): String = if (str == "") "" else str + "\n"
+
+    val (header0, footer0) = cfg.customOutputWrapper
+    val header =
+      if (hasErrors) ""
+      else withNewLine(header0) + "(function(){'use strict';\n"
+    val footer =
+      if (hasErrors) ""
+      else "}).call(this);\n" + withNewLine(footer0)
+
+    val outputContent = if (hasErrors) "" else withNewLine(compiler.toSource)
```

Comment:
I think protecting each individual block here with `hasErrors` adds a lot of clutter. From the code, I believe this was only introduced since otherwise `toSource` throws. Consider dropping all of it. We can write anything if an error happens.